### PR TITLE
add Integration specs

### PR DESCRIPTION
**Pull request by rugwizangoga**
# Blog app - integration specs for views and fix n+1 problems

- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user.
- Before:
before
- After:
After

##  Integration specs

- Used Capybara to write integration tests for each view in your project.

- User index page:

```
I can see the username of all other users.
I can see the profile picture for each user.
I can see the number of posts each user has written.
When I click on a user, I am redirected to that user's show page.
```
- user show page:

```
I can see the user's profile picture.
I can see the user's username.
I can see the number of posts the user has written.
I can see the user's bio.
I can see the user's first 3 posts.
I can see a button that lets me view all of a user's posts.
When I click Show More button, it redirects me to the user's post's index page.
```
- User post index page:

```
I can see the user's profile picture.
I can see the user's username.
I can see the number of posts the user has written.
I can see some of the post's body.
I can see the first comments on a post.
I can see how many comments a post has.
I can see how many likes a post has.
I can see a section for pagination if there are more posts than fit on the view.
When I click on a post, it redirects me to that post's show page.
```
- Post show page:

```
I can see who wrote the post.
I can see how many comments it has.
I can see how many likes it has.
I can see the post body.
I can see the username of each commentor.
I can see the comment each commentor left.
```

### CODE REVIEW

File: spec/views/users/index.html.erb_spec.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    visit users_path
+  end
+
+  it 'displays the username, profile picture, and number of posts for each user' do
+    @users_with_posts.each do |user|
+      expect(page).to have_css('.user-details h3', text: user.name)
+      expect(page).to have_css('.photo img') # Assuming there's an img tag for the photo
+      expect(page).to have_css('.user-posts', text: "Number of Posts: #{user.posts.count}")
+    end
+  end
```
Comment (joel-kalema) [inline spec/views/users/index.html.erb_spec.rb:15]: - [ ] Good job team 🥇, however, we  forgot to add the last test for the User index page:
````
- When I click on a user, I am redirected to that user's show page.
````
Kindly make sure you added it as it is [required](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=When%20I%20click%20on%20a%20user%2C%20I%20am%20redirected%20to%20that%20user%27s%20show%20page.) 👍

File: spec/views/users/show.html.erb_spec.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+require 'rails_helper'
+
+RSpec.describe 'User Show Page', type: :feature do
+  before do
+    @user = create(:user, :with_posts)
+    @most_recent_posts = @user.posts.order(created_at: :desc).limit(3)
+  end
```
Comment (joel-kalema) [inline spec/views/users/show.html.erb_spec.rb:7]: - [ ] You have done an excellent job, however, you forgot also to add tests for the user show page:
````
- I can see a button that lets me view all of a user's posts.
- When I click a user's post, it redirects me to that post's show page.
````
Kindly add them as it is [required](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#:~:text=I%20can%20see%20a%20button,to%20that%20post%27s%20show%20page.) 😄

File: spec/views/posts/index.html.erb_spec.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+
+RSpec.feature 'User Post Index Page', type: :feature do
+  let(:user) { FactoryBot.create(:user) }
+
+  before do
```
Comment (joel-kalema) [inline spec/views/posts/index.html.erb_spec.rb:6]: - [ ] Kindly make sure you add the test for:
````
- I can see a post's title.
- I can see the first comments on a post.
- I can see how many likes a post has.
- I can see a section for pagination if there are more posts than fit on the view.
````

**Review by joel-kalema — CHANGES_REQUESTED** [review]
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi Team
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

0bd223a2b156e0eaf61a713e84c3d3b7



## To highlight :champagne: 

- Tests are passing :heavy_check_mark: 
Screenshot 2023-08-11 122258

- Good PR description :heavy_check_mark: 
- Follow GitHub flow :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me @joel-kalema  when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/users/show.html.erb_spec.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  context 'Interactions' do
+    before { visit user_path(@user) }
+
+    it 'redirects to the user\'s posts index page when clicking "Show More" button' do
+      click_link 'Show More'
+      expect(page).to have_current_path(user_posts_path(@user))
+    end
+  end
```
Comment (rugwizangoga) [inline spec/views/users/show.html.erb_spec.rb]: @joel-kalema ,
thank you for the review but this requested added test were implemented above.
this test to be specific:
`- I can see a button that lets me view all of a user's posts.`

Comment (joel-kalema) [reply spec/views/users/show.html.erb_spec.rb]: @rugwizangoga 
Keep doing good job 🥇

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi 👋 @rugwizangoga, @Daniel-Kigozi, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

👍 Great job adding other tests as requested by the previous reviewer.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**